### PR TITLE
Attempt at modularising/improving code legibility

### LINKED ISSUE #41: Modularise input parameters?
The many input parameters to our entry points (e.g. `detect.py`) make the code hard to read, and the parameters hard to validate.
Would we benefit from creating some "structs" (e.g. one each for classification, detection, training) that may even provide simple functionality (like `get_defaults` ?) that could be validated individually? (see also #37 ).

### PR DESCRIPTION
**Pull request by alessandrofelder**
## Modularise "detect" widget
This PR reduces the number of arguments passed around to various functions by grouping related input parameters into
The number of nesting levels of function definitions is also reduced by this PR.
Only the detection widget is tackled in this PR.

This is related to #41 but I suggest we keep these open as long as we haven't modularised the rest of the code base too.
The magicgui warning (#34) is also removed as part of this, for the detection widget, but the problem persists in the training widget.

This PR might provide a blueprint to tackle modularisation elsewhere in the code.
I've also ran `black` and `isort` but I suspect formatting is still imperfect - I suggest we tackle this part separately as agreed (especially mypy, other linting checks seem to pass in the CI system).

The overall structure design has had a few iterations (commit history isn't great, but hey-ho), but improvement suggestion by the reviewers are welcome.

### CODE REVIEW

**Comment by paddyroddy** [comment]
I've just ran `pre-commit` to make the review easier. See https://pre-commit.com/#automatically-enabling-pre-commit-on-repositories to enable it always

File: cellfinder_napari/detect.py
Comment (paddyroddy) [inline cellfinder_napari/detect.py:1]: I'm curious, I've never found a need to use `math` over `numpy` unless trying to cut down on dependencies. Is that the idea here?

File: cellfinder_napari/detect.py
Comment (paddyroddy) [inline cellfinder_napari/detect.py:9]: has this been taken care of?

File: cellfinder_napari/detect.py
Comment (paddyroddy) [inline cellfinder_napari/detect.py:11]: again here?

File: cellfinder_napari/detect_utils.py
Comment (paddyroddy) [inline cellfinder_napari/detect_utils.py:18]: I'm assuming this is a `dict` and not a `tuple[dict]`?
```suggestion
    # some elements need to be remembered without having to create a new instance, for reset button
    persistent_defaults = {}
```

File: cellfinder_napari/detect_utils.py
Comment (paddyroddy) [inline cellfinder_napari/detect_utils.py:15]:
```suggestion
```

File: cellfinder_napari/detect_utils.py
Comment (paddyroddy) [inline cellfinder_napari/detect_utils.py:22]: I think we should really try to include typing as much as possible. Can be a lifesaver.

We could also add `mypy` to the `pre-commit` hook. What do you think?

File: cellfinder_napari/detect_utils.py
Comment (paddyroddy) [inline cellfinder_napari/detect_utils.py:29]: Are we being consistent with capitalisation? i.e. capital just for first word rather than all

```suggestion
     label = key.replace("_", " ").capitalize() if custom_label is None else custom_label
```

File: cellfinder_napari/detect_utils.py
Comment (paddyroddy) [inline cellfinder_napari/detect_utils.py:51]: this might be your design preference but I find the first blank line a bit odd 🤷

File: cellfinder_napari/detect_utils.py
Comment (paddyroddy) [inline cellfinder_napari/detect_utils.py:70]: This will have been caused by `black` but I think it looks better with comment at top

```suggestion
        # doesn't affect self, because asdict creates a copy.
        del data_input_dict["voxel_size_z"]  
```

File: cellfinder_napari/detect_utils.py
Comment (paddyroddy) [inline cellfinder_napari/detect_utils.py:98]: does this modify the parent `persistent_defaults` or create a new one?

What are the benefits of this over initialising the `dataclass` with the appropriate defaults

File: cellfinder_napari/detect_utils.py
Comment (paddyroddy) [inline cellfinder_napari/detect_utils.py:177]: does this need to be here?

File: cellfinder_napari/detect.py
Comment (paddyroddy) [inline cellfinder_napari/detect.py:30]: glad you moved the imports 🤮

File: cellfinder_napari/detect.py
Comment (paddyroddy) [inline cellfinder_napari/detect.py:32]: is this correct? Looks odd that the `img` tag isn't closed, but it might just be how this works

File: cellfinder_napari/detect.py
Comment (paddyroddy) [inline cellfinder_napari/detect.py:106]: In `python>=3.9`, potentially don't change now but this is the future! 🚀 

```suggestion
    ) -> list[napari.types.LayerDataTuple]:
```

File: cellfinder_napari/detect.py
Comment (paddyroddy) [inline cellfinder_napari/detect.py:152]: I'm taking your word for it that the number of inputs in the docstring matches the number in the function

**Review by paddyroddy — APPROVED** [review]
Looks good, but not tested it in practice

my main concern (which we discussed in person) is the fact that functions/classes are all in the same file etc

File: cellfinder_napari/detect_utils.py
Comment (adamltyson) [reply cellfinder_napari/detect_utils.py:22]: 100%. I'm (slowly) introducing it to other BG tools.

File: cellfinder_napari/detect.py
Comment (adamltyson) [reply cellfinder_napari/detect.py:1]: I've always just used in built functions over imported libraries. No real reason other than simplicity.

File: cellfinder_napari/detect.py
Comment (adamltyson) [reply cellfinder_napari/detect.py:174]: Curious to know people's thoughts on this. @paddyroddy's suggestion is probably more Pythonic, but I find it easier to parse the original.

Comment (paddyroddy) [reply cellfinder_napari/detect.py:174]: I didn't like it at first but `sourcery.ai` (which I used to use prior to having an M1 mac as it doesn't work) always used to suggest this to me and I grew to like it https://docs.sourcery.ai/refactorings/assign-if-exp/

File: cellfinder_napari/detect.py
Comment (adamltyson) [reply cellfinder_napari/detect.py:32]: Doubt it, probably me just chucking rubbish html into the repo.

File: cellfinder_napari/detect_utils.py
Comment (paddyroddy) [reply cellfinder_napari/detect_utils.py:22]: I'm going to raise a PR including this soon

**Comment by paddyroddy** [comment]
I've added a new CI for `pre-commit`, would be good if this could pass it before merging (but that is easier than it sounds) https://github.com/brainglobe/cellfinder-napari/pull/58

**Review by adamltyson — APPROVED** [review]

File: cellfinder_napari/detect.py
Comment (adamltyson) [reply cellfinder_napari/detect.py:30]: I just saw on zulip, this has changed, so the imports can be moved to the proper place.

**Comment by adamltyson** [comment]
FYI I've disabled the sonar cloud summary bot in a bid to reduce the amount of spam from PRs in this repo.

**Comment by paddyroddy** [comment]
> FYI I've disabled the sonar cloud summary bot in a bid to reduce the amount of spam from PRs in this repo.

thank you! 😆

File: cellfinder_napari/detect.py
Comment (alessandrofelder) [reply cellfinder_napari/detect.py:11]: Made an issue for this: #67 .

File: cellfinder_napari/detect_utils.py
Comment (alessandrofelder) [reply cellfinder_napari/detect_utils.py:18]: Yup - good catch. I've refactored and simplified this, so this now looks completely different..

File: cellfinder_napari/detect_utils.py
Comment (alessandrofelder) [reply cellfinder_napari/detect_utils.py:15]: Formatting to be taken care of in separate PR, as discussed.

File: cellfinder_napari/detect_utils.py
Comment (alessandrofelder) [reply cellfinder_napari/detect_utils.py:29]: Yes re capitalisation (at least that was the principle before this PR IIUC, and this PR should not change the UI.
I've added your suggestion here - thanks.

File: cellfinder_napari/detect_utils.py
Comment (alessandrofelder) [reply cellfinder_napari/detect_utils.py:70]: Fair point re comment. As discussed, the reason for this is because `cellfinder` core takes different kwargs than the widgets, so this function is here to facilitate this.

File: cellfinder_napari/detect_utils.py
Comment (alessandrofelder) [reply cellfinder_napari/detect_utils.py:177]: Nope, removed (We can add it back in (not as a comment) if useful in the future).
@adamltyson should having the classification batch size as an input be raised as an issue? Or might it be part of #67?

File: cellfinder_napari/detect.py
Comment (alessandrofelder) [reply cellfinder_napari/detect.py:50]: Good idea - I've simplified this a bit further. I suggest we leave further simplifications to separate PR as this is holding up some other work we'd like to do?

File: cellfinder_napari/detect.py
Comment (alessandrofelder) [reply cellfinder_napari/detect.py:107]: Not sure we can improve on this here, because AFAIK napari needs native types listed here to decide what widget to expose. But I could be wrong?

I think I've now managed to reduce the number of arguments that functions take everywhere else, by using the input container classes.

File: cellfinder_napari/detect_utils.py
Comment (adamltyson) [reply cellfinder_napari/detect_utils.py:177]: Eventually part of #67, but I have a feeling its commented out because the parameter doesn't work, even when passed to `cellfinder-core`.

File: cellfinder_napari/detect.py
Comment (alessandrofelder) [reply cellfinder_napari/detect.py:106]: Yup!

File: cellfinder_napari/detect.py
Comment (alessandrofelder) [reply cellfinder_napari/detect.py:152]: Yea... opening the plugin in the GUI doesn't work if not.

File: cellfinder_napari/detect.py
Comment (alessandrofelder) [reply cellfinder_napari/detect.py:9]: see #52.

File: cellfinder_napari/detect_utils.py
Comment (alessandrofelder) [reply cellfinder_napari/detect_utils.py:51]: Wasn't there on purpose 😅 Hope I've removed occurrences of this now (and otherwise can be part of formatting PR).

File: cellfinder_napari/detect_utils.py
Comment (alessandrofelder) [reply cellfinder_napari/detect_utils.py:98]: As discussed, I think this creates a new one. But yea, good point. I've changed to initialising a fresh dataclass and its defaults in the interest of avoiding code duplication.

File: cellfinder_napari/detect.py
Comment (alessandrofelder) [reply cellfinder_napari/detect.py:174]: It took getting used to for me too - equally easy for me now though, so I don't have a strong opinion either way - weakly leaning towards Paddy's suggestion.

**Review by paddyroddy — APPROVED** [review]
Looks good

**Pull request merged**
